File: server/src/routes/agentChat.js
import express from 'express';
import { SupervisorAgent } from '../agents/supervisorAgent.js';
import pool, { DEFAULT_USER_ID } from '../db.js';
import { memoryAgent } from '../agents/memoryAgent.js';
import { detectDocumentRequest, buildSummary } from '../services/documentTypeDetector.js';
import logger from '../utils/logger.js';
import { fetchChunksForChat, fetchImagesForChat } from '../rag.js';
import { classifyTask } from '../services/taskRouter.js';
import { calculateAtsScore, formatAtsAnswer } from '../services/atsService.js';
import { DocumentAnalysisAgent } from '../agents/documentAnalysisAgent.js';
import { DocumentComparisonAgent } from '../agents/documentComparisonAgent.js';
import { ResumeAnalysisAgent } from '../agents/resumeAnalysisAgent.js';
import { VisionAgent } from '../agents/visionAgent.js';
import { CodeAgent } from '../agents/codeAgent.js';

const router = express.Router();
const supervisor = new SupervisorAgent();
const documentAnalysisAgent = new DocumentAnalysisAgent();
const documentComparisonAgent = new DocumentComparisonAgent();
const resumeAnalysisAgent = new ResumeAnalysisAgent();
const visionAgent = new VisionAgent();
const codeAgent = new CodeAgent();

const STAGE_LABELS = {
  planning: 'Analyzing your question...',
  memory: 'Checking earlier context...',
  researching: 'Gathering evidence from sources...',
  reasoning: 'Reasoning through the evidence...',
  reviewing: 'Reviewing the answer for accuracy...',
  regenerating: 'Found an issue — revising the answer...'
};

// Critical-thinking pipeline endpoint: streams stage progress ("Planning...",
// "Researching...", ...) while SupervisorAgent runs its full
// plan -> research -> reason -> review loop (no token streaming mid-answer —
// the answer only exists once reasoning+review finish, unlike the simple
// single-shot /api/chats/:chatId/query path). This trades the 60s timeout
// budget of that endpoint for minutes of real deliberation, intentionally,
// per the "Reasoning > Speed" priority for this pipeline.
router.post('/', async (req, res) => {
  const { chatId, message } = req.body;

  if (!chatId || !message) {
    return res.status(400).json({ error: 'chatId and message are required' });
  }

  res.setHeader('Content-Type', 'text/event-stream');
  res.setHeader('Cache-Control', 'no-cache');
  res.setHeader('Connection', 'keep-alive');
  res.flushHeaders();

  const reqId = `${chatId}-${Date.now()}`;
  console.log(`[CRITICAL_THINK_START] ${reqId}`);

  try {
    const userMsgResult = await pool.query(
      'INSERT INTO messages (chat_id, role, content) VALUES ($1, $2, $3) RETURNING *',
      [chatId, 'user', message]
    );
    const userMsg = userMsgResult.rows[0];
    memoryAgent.extractMemory(DEFAULT_USER_ID, chatId, message).catch(() => {});

    // Task Router applies here too — "analyze/evaluate" keywords can route
    // a message to this deep pipeline (see Chat.jsx's complexity heuristic),
    // but an attached file still means Uploaded Document > Web Search. Same
    // short-circuit as the fast path, before SupervisorAgent (and its
    // ResearchAgent web search) ever runs.
    const [allChatChunks, chatImages] = await Promise.all([
      fetchChunksForChat(chatId),
      fetchImagesForChat(chatId)
    ]);
    const hasFiles = allChatChunks.length > 0;
    const hasImages = chatImages.length > 0;
    const fileCount = new Set(allChatChunks.map(c => c.file_id)).size;
    const task = classifyTask(message, { hasFiles, hasImages, fileCount });
    console.log(`[TASK_ROUTE] ${reqId} type=${task.type} hasFiles=${hasFiles} hasImages=${hasImages}`);

    if (task.type === 'vision') {
      const image = chatImages[chatImages.length - 1];
      const result = await visionAgent.run(message, { filePath: image.file_path, fileName: image.original_filename });
      const answer = result.output.answer;
      const confidence = { score: result.success ? 75 : 0, label: result.success ? 'high' : 'low', reason: 'Based on uploaded image content only.' };

      await pool.query('INSERT INTO messages (chat_id, role, content) VALUES ($1, $2, $3)', [chatId, 'ai', answer]);
      await pool.query('UPDATE chats SET updated_at = CURRENT_TIMESTAMP WHERE id = $1', [chatId]);
      res.write(`data: ${JSON.stringify({ text: answer, sources: [{ filename: image.original_filename, type: 'image' }], confidence })}\n\n`);
      res.write('data: [DONE]\n\n');
      return res.end();
    }

    if (task.type === 'coding') {
      // Coding questions never belong in the 1-3min critical-thinking
      // pipeline (planner->research->reasoning->review) — they're
      // self-contained, the research stages add latency with no accuracy
      // benefit. Same fast path as the single-shot endpoint.
      let codeAnswer = '';
      try {
        const { answer, confidence } = await codeAgent.runStream(message, (token) => {
          codeAnswer += token;
          res.write(`data: ${JSON.stringify({ text: codeAnswer, sources: [] })}\n\n`);
        });
        codeAnswer = answer;
        await pool.query('INSERT INTO messages (chat_id, role, content) VALUES ($1, $2, $3)', [chatId, 'ai', codeAnswer]);
        await pool.query('UPDATE chats SET updated_at = CURRENT_TIMESTAMP WHERE id = $1', [chatId]);
        res.write(`data: ${JSON.stringify({ text: codeAnswer, sources: [], confidence })}\n\n`);
      } catch (err) {
        res.write(`data: ${JSON.stringify({ error: err.message })}\n\n`);
      }
      res.write('data: [DONE]\n\n');
      return res.end();
    }

    if (task.type !== 'general') {
      const documentText = allChatChunks.map(c => c.content).join('\n');
      const fileName = allChatChunks[0]?.original_filename || 'document';
      let answer, confidence;

      if (task.type === 'document_comparison') {
        const byFile = new Map();
        for (const chunk of allChatChunks) {
          const existing = byFile.get(chunk.file_id) || { fileName: chunk.original_filename, text: '' };
          existing.text += `${chunk.content}\n`;
          byFile.set(chunk.file_id, existing);
        }
        const documents = [...byFile.values()];
        const result = await documentComparisonAgent.run(message, { documents });
        answer = result.output.answer;
        confidence = { score: result.success ? 75 : 0, label: result.success ? 'high' : 'low', reason: 'Based on the uploaded documents only.' };
        const sourceCards = documents.map(d => ({ filename: d.fileName, type: 'document' }));

        await pool.query('INSERT INTO messages (chat_id, role, content) VALUES ($1, $2, $3)', [chatId, 'ai', answer]);
        await pool.query('UPDATE chats SET updated_at = CURRENT_TIMESTAMP WHERE id = $1', [chatId]);
        res.write(`data: ${JSON.stringify({ text: answer, sources: sourceCards, confidence })}\n\n`);
        res.write('data: [DONE]\n\n');
        return res.end();
      }

      if (task.type === 'ats') {
        const atsResult = calculateAtsScore(documentText);
        answer = formatAtsAnswer(atsResult);
        confidence = { score: atsResult.score, label: atsResult.score >= 70 ? 'high' : atsResult.score >= 30 ? 'medium' : 'low', reason: 'Calculated from parsed resume content.' };
      } else if (task.type === 'resume_analysis') {
        const result = await resumeAnalysisAgent.run(message, { documentText, fileName });
        answer = result.output.answer;
        confidence = { score: result.success ? 75 : 0, label: result.success ? 'high' : 'low', reason: 'Based on uploaded resume content only.' };
      } else {
        const result = await documentAnalysisAgent.run(message, { documentText, fileName });
        answer = result.output.answer;
        confidence = { score: result.success ? 75 : 0, label: result.success ? 'high' : 'low', reason: 'Based on uploaded document content only.' };
      }

      await pool.query('INSERT INTO messages (chat_id, role, content) VALUES ($1, $2, $3)', [chatId, 'ai', answer]);
      await pool.query('UPDATE chats SET updated_at = CURRENT_TIMESTAMP WHERE id = $1', [chatId]);
      res.write(`data: ${JSON.stringify({ text: answer, sources: [{ filename: fileName, type: 'document' }], confidence })}\n\n`);
      res.write('data: [DONE]\n\n');
      return res.end();
    }

    const onStage = (stage) => {
      console.log(`[STAGE] ${reqId} ${stage}`);
      res.write(`data: ${JSON.stringify({ stage, stageLabel: STAGE_LABELS[stage] || stage })}\n\n`);
    };

    const result = await supervisor.run(message, { chatId, excludeMessageId: userMsg.id, onStage, hasFiles });

    if (!result.success) {
      throw new Error(result.error || 'Agent execution failed');
    }

    const { answer, evidence, confidence, contradictions, regenerated } = result.output;

    const docType = detectDocumentRequest(message);
    const document = docType ? {
      title: docType.label,
      subtitle: null,
      type: docType.type,
      summary: buildSummary(answer),
      content: answer,
      createdAt: new Date().toISOString(),
      exportFormats: ['docx', 'pdf', 'pptx', 'xlsx', 'markdown', 'txt']
    } : null;

    await pool.query(
      'INSERT INTO messages (chat_id, role, content, document) VALUES ($1, $2, $3, $4)',
      [chatId, 'ai', answer, document ? JSON.stringify(document) : null]
    );
    await pool.query('UPDATE chats SET updated_at = CURRENT_TIMESTAMP WHERE id = $1', [chatId]);

    console.log(`[CRITICAL_THINK_DONE] ${reqId} regenerated=${regenerated} confidence=${confidence.score}`);

    res.write(`data: ${JSON.stringify({ text: answer, sources: evidence, confidence, contradictions, document })}\n\n`);
    res.write('data: [DONE]\n\n');
    res.end();
  } catch (err) {
    logger.error('agentChat.failed', { chatId, error: err.message });
    console.error(`[CRITICAL_THINK_ERROR] ${reqId}`, err.message);
    res.write(`data: ${JSON.stringify({ error: err.message })}\n\n`);
    res.write('data: [DONE]\n\n');
    res.end();
  }
});

export default router;

File: server/app/routes/agent_chat.py
import asyncio
import json
import time
from datetime import datetime, timezone

from fastapi import APIRouter, Depends, HTTPException
from fastapi.responses import StreamingResponse
from pydantic import BaseModel

from ..agents.code_agent import CodeAgent
from ..agents.memory_agent import memory_agent
from ..agents.tool_agent import ToolAgent
from ..core import db
from ..core.logger import logger
from ..middleware.auth_middleware import get_current_user
from ..rag import fetch_chunks_for_chat, fetch_images_for_chat
from ..services.document_type_detector import detect_document_request, build_summary
from ..services.task_router import classify_task
from ..utils.stream_cancel import drain_task_queue

router = APIRouter()

tool_agent = ToolAgent()
code_agent = CodeAgent()

STAGE_LABELS = {
    "planning": "Analyzing your question...",
    "memory": "Checking earlier context...",
    "researching": "Gathering evidence from sources...",
    "reasoning": "Reasoning through the evidence...",
    "reviewing": "Reviewing the answer for accuracy...",
    "regenerating": "Found an issue — revising the answer...",
}


def _sse(data: dict) -> str:
    return f"data: {json.dumps(data)}\n\n"


class AgentChatBody(BaseModel):
    chatId: str | None = None
    message: str | None = None


@router.post("/api/agent/chat")
async def agent_chat(body: AgentChatBody, current_user: dict = Depends(get_current_user)):
    if not body.chatId or not body.message:
        raise HTTPException(status_code=400, detail="chatId and message are required")

    chat_id = body.chatId
    message = body.message
    user_id = current_user["id"]

    chat_check = await db.query("SELECT id FROM chats WHERE id = $1 AND user_id = $2", [chat_id, user_id])
    if chat_check["rowCount"] == 0:
        raise HTTPException(status_code=404, detail="Chat not found")

    async def stream():
        req_id = f"{chat_id}-{int(time.time() * 1000)}"
        try:
            user_msg_result = await db.query(
                "INSERT INTO messages (chat_id, role, content) VALUES ($1, $2, $3) RETURNING *",
                [chat_id, "user", message],
            )
            user_msg = user_msg_result["rows"][0]
            try:
                await memory_agent.extract_memory(user_id, chat_id, message)
            except Exception:
                pass

            all_chat_chunks = await fetch_chunks_for_chat(chat_id)
            chat_images = await fetch_images_for_chat(chat_id)
            has_files = len(all_chat_chunks) > 0
            has_images = len(chat_images) > 0
            file_count = len({c["file_id"] for c in all_chat_chunks})
            task = classify_task(message, has_files=has_files, has_images=has_images, file_count=file_count)
            logger.info("task.route", {"reqId": req_id, "type": task["type"], "hasFiles": has_files, "hasImages": has_images})

            if task["type"] == "coding":
                try:
                    token_queue: asyncio.Queue = asyncio.Queue()
                    code_answer_parts = []

                    def on_token(token):
                        code_answer_parts.append(token)
                        token_queue.put_nowait("".join(code_answer_parts))

                    gen_task = asyncio.create_task(code_agent.run_stream(message, on_token))

                    # Cancels gen_task if the client aborts mid-stream.
                    async for text in drain_task_queue(gen_task, token_queue):
                        yield _sse({"text": text, "sources": []})

                    result = await gen_task
                    code_answer = result["answer"]
                    confidence = result["confidence"]
                    await db.query("INSERT INTO messages (chat_id, role, content) VALUES ($1, $2, $3)", [chat_id, "ai", code_answer])
                    await db.query("UPDATE chats SET updated_at = CURRENT_TIMESTAMP WHERE id = $1", [chat_id])
                    yield _sse({"text": code_answer, "sources": [], "confidence": confidence})
                except Exception as err:
                    yield _sse({"error": str(err)})
                yield "data: [DONE]\n\n"
                return

            stage_queue: asyncio.Queue = asyncio.Queue()

            def on_stage(stage):
                stage_queue.put_nowait({"stage": stage, "stageLabel": STAGE_LABELS.get(stage, stage)})

            tool_agent_task = asyncio.create_task(
                tool_agent.run(message, {"chatId": chat_id, "excludeMessageId": user_msg["id"], "onStage": on_stage, "hasFiles": has_files})
            )

            # Cancels the whole multi-agent run if the client aborts.
            async for event in drain_task_queue(tool_agent_task, stage_queue, poll_interval=0.2):
                yield _sse(event)

            result = await tool_agent_task

            if not result["success"]:
                raise RuntimeError(result.get("error") or "Agent execution failed")

            output = result["output"]
            answer, evidence, confidence, contradictions = output["answer"], output["evidence"], output["confidence"], output["contradictions"]

            doc_type = detect_document_request(message)
            document = None
            if doc_type:
                document = {
                    "title": doc_type["label"],
                    "subtitle": None,
                    "type": doc_type["type"],
                    "summary": build_summary(answer),
                    "content": answer,
                    "createdAt": datetime.now(timezone.utc).isoformat(),
                    "exportFormats": ["docx", "pdf", "pptx", "xlsx", "markdown", "txt"],
                }

            await db.query(
                "INSERT INTO messages (chat_id, role, content, document) VALUES ($1, $2, $3, $4)",
                [chat_id, "ai", answer, json.dumps(document) if document else None],
            )
            await db.query("UPDATE chats SET updated_at = CURRENT_TIMESTAMP WHERE id = $1", [chat_id])

            yield _sse({"text": answer, "sources": evidence, "confidence": confidence, "contradictions": contradictions, "document": document})
            yield "data: [DONE]\n\n"
        except Exception as err:
            logger.error("agentChat.failed", {"chatId": chat_id, "error": str(err)})
            yield _sse({"error": str(err)})
            yield "data: [DONE]\n\n"

    return StreamingResponse(stream(), media_type="text/event-stream")
